### ISSUE TITLE
CI: windows-latest cells die with 0xc000001d in the NF4 training-step test — three occurrences, now on main

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#382]
## Context

`tests/test_v07202.py::TestNF4EndToEndSetup::test_one_nf4_training_step_actually_runs` runs a real training step through bitsandbytes' 4-bit kernels. On some `windows-latest` runners the wheel executes an instruction the CPU does not have, and the process dies with `Windows fatal exception: code 0xc000001d` (ILLEGAL_INSTRUCTION) — a faulthandler stack dump with no Python exception.

This is already documented in `.claude/CLAUDE.md` (release checklist step 16) as "a runner/wheel problem, not a code regression — re-run before diagnosing". That guidance is correct. What has changed is the frequency, and that it now lands on `main`.

## Evidence

| when | commit | cell | outcome |
|---|---|---|---|
| 2026-08-05 | `9117da1` (two lines of markdown) | windows-latest / **3.12** | crash; `gh run rerun --failed` passed on the same tree |
| 2026-08-11 | PR #362 [`31466333869`](https://github.com/MakazhanAlpamys/Soup/actions/runs/31466333869) | windows-latest / **3.10** | crash; rerun passed, 11/11 |
| 2026-08-11 | `7c9a931` (the #362 merge) [`31468811428`](https://github.com/MakazhanAlpamys/Soup/actions/runs/31468811428) | windows-latest / **3.10** | crash |

Three occurrences, two Python versions, so it tracks the **runner CPU**, not the interpreter. GitHub's `windows-latest` fleet is heterogeneous and a share of it lacks whatever instruction the bitsandbytes wheel emits.

The strongest control is free: `1715261` is `7c9a931` **plus one line in `CONTRIBUTORS.md`** and its run came back green on the same pool. So the failing tree and the passing tree differ by a docs line.

Both crashes land at the same point — the last line before the dump is always `test_setup_builds_a_real_trl_trainer_under_nf4 PASSED`, i.e. the process dies inside the very next test, which is the first one to call `trainer.train()`.

## Why it is worth acting on rather than re-running

An illegal instruction is not a test failure — it kills the interpreter, so the cell reports **nothing** about the other ~17,000 tests it had not reached yet (the crash lands at 97%, but that is luck, not design). A red `main` that means "one binary wheel met the wrong CPU" also trains the reader to skip past red, which is the expensive part.

`test_the_saved_adapter_is_canonical`, two tests below, calls `trainer.train()` as well and would crash identically; it survives only because the process is already gone.

## Proposed fix path

Mirror the file's own `skip_on_mps` idiom — that marker exists for exactly this shape of problem (a real training step on a platform whose kernels are not supported, skipped rather than making an unverified claim).

Skip **in CI only**, not on Windows generally:

```python
skip_on_windows_ci = pytest.mark.skipif(
    sys.platform == "win32" and os.environ.get("CI") == "true",
    reason=(
        "part of GitHub's windows-latest fleet lacks an instruction the "
        "bitsandbytes wheel emits; the 4-bit kernel dies with 0xc000001d, "
        "which kills the interpreter and reports nothing about the rest of "
        "the suite. Covered on ubuntu + macos, and still live on a developer's "
        "Windows box, where the CPU is known."
    ),
)
```

Applied to the two tests in `TestNF4EndToEndSetup` that call `trainer.train()`. The condition is deliberately narrow: the maintainer develops on Windows, so the test must keep running there — what is being skipped is an unknown CPU, not a platform.

Coverage is not lost: the same tests run on ubuntu 3.10/3.11/3.12 and macos 3.10/3.11/3.12, and the assertion (`pool.loads > 0`) is platform-independent.

## Alternatives considered

- **Keep re-running.** Works, and is the current documented policy. Costs a full 17k-test cell and a manual `gh run rerun --failed` every time, and leaves red commits in `main`'s history that are indistinguishable at a glance from real ones.
- **Pin a bitsandbytes version.** Does not address it — the wheel is fine, the runner CPU is the variable.
- **Catch the fault.** Not possible; `SIGILL` on Windows is not deliverable to Python.

## Acceptance criteria

- [ ] `windows-latest` cells no longer crash the interpreter on the NF4 training-step tests
- [ ] The same tests still run — and can still fail — on ubuntu and macos, and on a local Windows box
- [ ] The skip reason names the actual cause, so nobody later reads it as "NF4 does not work on Windows"
- [ ] `.claude/CLAUDE.md`'s step-16 note is updated to point here rather than prescribing a re-run

**Comment by MakazhanAlpamys** [comment]
Implemented in [`5329bee`](https://github.com/MakazhanAlpamys/Soup/commit/5329bee) — level 1 only, tests-only, no `src/` change and no version bump.

`_windows_ci()` is `sys.platform == "win32" and os.environ.get("CI") == "true"`, guarding the two tests in `TestNF4EndToEndSetup` that reach a real `trainer.train()`. Narrow on purpose: GitHub Actions sets `CI`, nothing on a developer's machine does, so a local Windows box keeps running them — what is excluded is an unknown CPU, not a platform.

Verified both directions on Windows before pushing:

```
pytest tests/test_v07202.py            -> 92 passed          (no CI var: the step really runs)
CI=true pytest tests/test_v07202.py    -> 90 passed, 2 skipped, reason named
```

`TestTheWindowsCiSkipStaysNarrow` pins all four edges (windows+CI fires; windows without CI does not; linux+CI and macos+CI do not). That class is the load-bearing part: a skipif that quietly widened would remove the NF4 training step everywhere and nothing would say so, since a skipped test and a passing test are the same colour — a worse silent failure than the one being fixed.

Acceptance criteria 1-3 are met. The fourth (CLAUDE.md's step-16 note) is done locally — that file is gitignored, so it cannot be referenced here; it now points at this issue and says *not* to re-run, since seeing the signature again means either the guard drifted or a new test reached the kernel.

Leaving this open until a few windows-latest cells have run green on `main`, then closing.

Level 2 — running that one test in a subprocess so an illegal instruction becomes a reportable failure instead of a dead cell — is the principled way to get the Windows coverage back, and is deliberately not done here. Worth its own issue if the coverage is ever wanted.

**Comment by MakazhanAlpamys** [comment]
Closing: three consecutive `main` runs since [`5329bee`](https://github.com/MakazhanAlpamys/Soup/commit/5329bee) have all nine cells green, including `windows-latest` 3.10 / 3.11 / 3.12 — the three that were the coin flip.

| run | commit | windows cells |
|---|---|---|
| [31473693701](https://github.com/MakazhanAlpamys/Soup/actions/runs/31473693701) | `5329bee` | 3.10 / 3.11 / 3.12 success |
| [31476719741](https://github.com/MakazhanAlpamys/Soup/actions/runs/31476719741) | `fccd586` | 3.10 / 3.11 / 3.12 success |

Acceptance criteria 1-3 met and verified; the fourth (the step-16 note in `.claude/CLAUDE.md`) is done locally — that file is gitignored, so it cannot be linked here. It now points at this issue and says explicitly **not** to re-run, because seeing this signature again means either the guard drifted or a new test reached the bitsandbytes kernel.

Level 2 — running that one test in a subprocess so an illegal instruction becomes a reportable failure rather than a dead cell — is not done, deliberately. It is the principled way to get the Windows coverage back and it is worth its own issue if that coverage is ever wanted; today the coverage lives on ubuntu ×3 and macos ×3, and on any Windows box whose CPU is known.

**Comment by Srinivasan8888** [comment]
Two 2026-09-17 occurrences that do not fit the recorded diagnosis, moved here from #1062 so they are not lost in a PR thread.

**The runs**, both on PRs that touch nothing related:

| run | cell | died in |
|---|---|---|
| 35228388681 | `test (windows-latest, 3.12)` | `tests/test_rewind_hf.py::test_recorded_rows_match_sampler_order` |
| 35254266989 | `test (windows-latest, 3.11)` | the same test |

Same signature both times: `Windows fatal exception: code 0xc000001d`, a faulthandler dump, no Python exception.

**Why they do not fit.** `tests/_windows_ci.py`'s docstring records the cause as an instruction the **bitsandbytes** wheel emits. But the crash site does not load bitsandbytes: `test_rewind_hf.py:82`'s `_trainer` builds the model with a plain `AutoModelForCausalLM.from_pretrained(MODEL_ID)` — no `quantization_config`, CPU — and `test_rewind_wiring.py`'s marked test goes through `test_issue341_seed_and_fullft._wrapper`, which sets `quantization: "none"`. Whatever emits the instruction on those runners, these two crashes did not come through bnb.

(Correcting myself: the first version of this comment said the crash site went through `_wrapper`. It does not; that is the wiring test. Neither path loads bitsandbytes, so the point stands.)

**A second fact from the same logs, which bounds it further.** On both cells the full fine-tune tests in `test_issue341_seed_and_fullft.py` — a real `trainer.train()` over the same tiny Llama, same helper — **passed at 12%**, long before the crash at 41%. So it is not "any real training step dies" either. The difference between the survivors and the crash site is the rewind recorder attached to the trainer, which does per-row loss work inside the step.

I am not claiming that is the cause; it is the narrowest statement the two logs support. Filing it because the docstring's bnb explanation would send the next person looking in the wrong place, and because #1059 and #1062 skip these tests rather than explain them.

Both python versions (3.11 and 3.12) and, per the existing record, 3.10 as well, so it still tracks the runner CPU rather than the interpreter.
